### DISCUSSION TITLE
Event delegation for parallel state nodes

### DISCUSSION THREAD

**Discussion by webstacker** [statelyai/xstate#1974] [Q&A]
Using generic event names across state nodes causes all states to transition given that event. This makes sense if the event is generic. Other than namespacing the events themselves e.g. `UPLOAD.CLICK`, is there any way to indicate where an event originated, so that the event can be delegated correctly? Similar concept to:

https://developer.mozilla.org/en-US/docs/Web/API/Event/target#example

It seems like separate machines would be required to model parallel state with shared event names.

Pseudo code example (note the transition event):

```JavaScript
const { Machine, assign } = require("xstate");

const fileMachine = Machine({
    id: "file",
    type: "parallel",
    states: {
        upload: {
            initial: "idle",
            states: {
                idle: {
                    on: {
                        CLICK: "pending"
                    }
                },
                pending: {
                    on: {
                        COMPLETE: "success"
                    }
                },
                success: {}
            },
        },
        download: {
            initial: "idle",
            states: {
                idle: {
                    on: {
                        CLICK: "pending"
                    }
                },
                pending: {
                    on: {
                        COMPLETE: "success"
                    }
                },
                success: {}
            },
        },
    },
});

console.log(fileMachine.initialState.value);
// => {
//   upload: 'idle',
//   download: 'idle'
// }

console.log(
    fileMachine.transition(
        {
            upload: "idle",
            download: "idle"
        },
        {
            event: "CLICK",
            target: "upload"
        }
    ).value
);
// => {
//   upload: 'pending',
//   download: 'idle'
// }
```

**Comment by Andarist** [comment]
As you have mentioned - you have two basic options here:
1. namespace events
2. split this into child machines 

Alternatively, you could try to add guards to your transitions and prevent a particular parallel region from transitioning based on some event data.

**Reply by webstacker** [reply]
Thanks @Andarist , I had considered guards too, but it seemed a bit complex for the problem.

**Comment by davidkpiano** [comment]
Making event names less generic guides you toward the pit of success; for example, `"CLICK"` doesn't mean anything, other than that "something was clicked", which isn't really descriptive.

However, `"DOWNLOAD.CLICK"` is more descriptive, and solves the issue.

You can also choose to add a parameter to your event that indicates its "source", but that's up to you to implement. Then you can make a helper like so:

```js
// Example implementation, do your own
const fromSource = (source, eventConfig) => ({
  ...eventConfig,
  cond: (_, event) => event.source === source
});

// ...

on: {
  CLICK: fromSource('download', { target: 'somewhere' })
}
```

**Reply by webstacker** [reply]
Thanks @davidkpiano , I might not be thinking about this correctly. I get they're solving different problems, but the DOM has a finite amount of events that can be applied to different things. I assumed that this should be possible here too. If I update my example to include your more descriptive event, it still seems like there is a place for "event targets"?

```JavaScript

const fileMachine = Machine({
    id: "file",
    type: "parallel",
    states: {
        thing1: {
            initial: "idle",
            states: {
                idle: {
                    on: {
                        "DOWNLOAD.CLICK": "pending"
                    }
                },
                pending: {
                    on: {
                        COMPLETE: "success"
                    }
                },
                success: {}
            },
        },
        thing2: {
            initial: "idle",
            states: {
                idle: {
                    on: {
                        "DOWNLOAD.CLICK": "pending"
                    }
                },
                pending: {
                    on: {
                        COMPLETE: "success"
                    }
                },
                success: {}
            }
        }
    }
});
```

Also, looking at this example https://xstate.js.org/docs/#parallel-state-machines. The "namespaced" event names are mirroring the "event target" functionality, but in a less explicit way. There is also a bit of duplication in the event naming / state name. One other thing is, how would you toggle all toggle'able things given they all have different events. Using shared events with an event target would give options such as:

```JavaScript
const nextState = wordMachine.transition(
    {
        bold: 'off',
        italics: 'off',
        underline: 'on',
        list: 'bullets'
    },
    'TOGGLE'
).value;
// {
//   bold: 'on',
//   italics: 'on',
//   underline: 'on',
//   list: 'bullets'
// }


const nextState = wordMachine.transition(
    {
        bold: 'off',
        italics: 'off',
        underline: 'on',
        list: 'bullets'
    },
    {
        event: 'TOGGLE',
        target: 'bold'
    }
).value;
// {
//   bold: 'on',
//   italics: 'off',
//   underline: 'on',
//   list: 'bullets'
// }
```

Is this the wrong way to be thinking about parallel state? Thanks

**Reply by Andarist** [reply]
> If I update my example to include your more descriptive event, it still seems like there is a place for "event targets"?

i dont quite get your point here - could u elaborate? Unless the part after the example is about this

> The "namespaced" event names are mirroring the "event target" functionality, but in a less explicit way

Yes, both can serve the same purpose. Would targets be more explicit? I wouldnt say so - events can have arbitrary properties, so adding a special one there isnt really that much explicit. One could also make a mistake and think that this is targeting a child machine (and not the child state).

> There is also a bit of duplication in the event naming / state name. 

This doesnt really feel much different from targets - either way u have to put ur targer somewhere and its always a „duplication”

> One other thing is, how would you toggle all toggle'able things given they all have different events.

Be specific and just have an additional TOGGLE_ALL event instead of trying to do multiple different things using a single event name.
